trackpy/find_link.py

```diff
@@ -25,60 +25,85 @@
 
     Empty frames will be returned as empty arrays of shape (0, ndim)."""
     ndim = len(pos_columns)
     grouped = iter(df.groupby(t_column))  # groupby sorts by default
 
     # get the first frame to learn first frame number
     cur_frame, frame = next(grouped)
-    next_frame = int(cur_frame) + 1
-    yield frame[pos_columns].values
+    cur_frame = int(cur_frame)
+    yield cur_frame, frame[pos_columns].values
+    cur_frame += 1
 
     for frame_no, frame in grouped:
-        while next_frame < int(frame_no):
-            next_frame += 1
-            yield np.empty((0, ndim))
+        frame_no = int(frame_no)
+        while cur_frame < frame_no:
+            yield cur_frame, np.empty((0, ndim))
+            cur_frame += 1
 
-        next_frame += 1
-        yield frame[pos_columns].values
+        yield cur_frame, frame[pos_columns].values
+        cur_frame += 1
 
 
-def link_simple_iter(coords_iter, search_range, memory=0):
+def coords_from_df_iter(df_iter, pos_columns, t_column):
+    """A generator that returns ndarrays of coords from a generator of
+    DataFrames. Also returns the first value of the t_column."""
+    ndim = len(pos_columns)
+
+    for df in df_iter:
+        if len(df) == 0:
+            yield None, np.empty((0, ndim))
+        else:
+            yield df[t_column].iloc[0], df[pos_columns].values
+
+
+def link_simple_iter(coords_iter, search_range, memory=0, predictor=None):
     """Link an iterable of per-frame coordinates into trajectories.
 
     Parameters
     ----------
-    coords_iter : iterable of 2d numpy arrays
+    coords_iter : iterable or enumerated iterable of 2d numpy arrays
     search_range : float or tuple
     memory : integer
+    predictor : predictor function; see 'predict' module
 
     Returns
     -------
     yields tuples (index, list of particle ids)
     """
     # ensure that coords_iter is iterable
     coords_iter = iter(coords_iter)
-    # take the first and obtain dimensionality
-    coords = next(coords_iter)
+
+    # interpret the first element of the iterable
+    val = next(coords_iter)
+    if isinstance(val, np.ndarray):
+        # the iterable was not enumerated, so enumerate the remainder
+        coords_iter = enumerate(coords_iter, start=1)
+        t, coords = 0, val
+    else:
+        t, coords = val
+
+    #  obtain dimensionality
     ndim = coords.shape[1]
     search_range = validate_tuple(search_range, ndim)
 
     # initialize the linker and yield the particle ids of the first frame
-    linker = Linker(search_range, memory)
-    linker.init_level(coords, t=0)
-    yield 0, linker.particle_ids
+    linker = Linker(search_range, memory, predictor=predictor)
+    linker.init_level(coords, t)
+    yield t, linker.particle_ids
 
-    for t, coords in enumerate(coords_iter, start=1):
+    for t, coords in coords_iter:
         linker.next_level(coords, t)
         for k, coord in enumerate(linker.coords):
```

Comment:
These two lines are obsolete (I recall to test whether the coords had been changed shuffled), they should be removed for better performance